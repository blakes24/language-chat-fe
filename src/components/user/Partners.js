import { useEffect } from "react";
import UserList from "../user/UserList";
import Container from "@material-ui/core/Container";
import Typography from "@material-ui/core/Typography";
import { makeStyles } from "@material-ui/core/styles";
import { useSelector, useDispatch } from "react-redux";
import { fetchPartners } from "../../store/partnerSlice";
import { Link } from "react-router-dom";
import Loading from "../Loading";

const useStyles = makeStyles({
  root: {
    marginTop: "1.5rem",
    marginBottom: ".5rem",
  },
  formControl: {
    marginBottom: ".5rem",
    minWidth: 240,
  },
  formContainer: {
    display: "flex",
    justifyContent: "center",
  },
  err: {
    textAlign: "center",
  },
});

function Partners() {
  const classes = useStyles();
  const partners = useSelector((state) => state.partners.items);
  const user = useSelector((state) => state.users.current);
  const loading = useSelector((state) => state.partners.loading);
  const error = useSelector((state) => state.partners.error);
  const dispatch = useDispatch();
  
  useEffect(() => {
    function getPartners() {
      try {
        dispatch(fetchPartners(user.id));
      } catch (err) {
        console.error(err);
      }
    }
    getPartners();
  }, [user, dispatch]);

  return (
    <Container>
      <Typography
        variant="h4"
        component="h1"
        align="center"
        className={classes.root}
      >
        Partners
      </Typography>

      {!partners.length && loading === "pending" && <Loading solid />}
      {error &&
        (Array.isArray(error) ? (
          error.map((err) => <p className={classes.err}>{err}</p>)
        ) : (
          <p className={classes.err}>{error.message}</p>
        ))}
      {partners.length > 0 ? (
        <UserList users={partners} partner={true} />
      ) : (
        <Typography align="center" className={classes.root}>
          No partners yet. Find users to chat with <Link to="/">here</Link>.
        </Typography>
      )}
    </Container>
  );
}
export default Partners;
