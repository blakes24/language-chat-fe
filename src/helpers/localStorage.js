export function setLocalStorage(key, val) {
  window.localStorage.setItem(key, JSON.stringify(val));
}

export function getFromLocalStorage(key, defaultValue) {
  let value;
  try {
    value = JSON.parse(
      window.localStorage.getItem(key) || JSON.stringify(defaultValue)
    );
  } catch (e) {
    console.error(e);
    value = defaultValue;
  }
  return value;
}
