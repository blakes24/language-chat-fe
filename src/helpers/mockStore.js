import configureStore from "redux-mock-store";

const middlewares = [];
const mockStore = configureStore(middlewares);

const testUser = {
  id: 1,
  name: "Test",
  bio: "stuff about mae",
  email: "test@mail.com",
  imageUrl: "https://cdn.fakercloud.com/avatars/happypeter1983_128.jpg",
  active: true,
  speaks: [
    {
      code: "it",
      language: "Italian",
    },
  ],
  learning: [
    {
      code: "ko",
      language: "Korean",
      level: "2",
    },
  ],
};

const initialState = {
  users: {
    current: testUser,
    items: [],
    token: "",
    loading: "idle",
    error: null,
  },
  rooms: { items: [], current: null, loading: "idle", error: null },
  messages: { items: [], loading: "idle", error: null },
  partners: { items: [], loading: "idle", error: null },
};

export const store = mockStore(initialState);
