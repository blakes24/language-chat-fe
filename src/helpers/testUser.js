import UserContext from "./UserContext";

const testUser = {
  id: 1,
  name: "Test",
  bio: "stuff about mae",
  email: "test@mail.com",
  imageUrl: "https://cdn.fakercloud.com/avatars/happypeter1983_128.jpg",
  active: true,
  speaks: [
    {
      code: "it",
      language: "Italian",
    },
  ],
  learning: [
    {
      code: "ko",
      language: "Korean",
      level: "2",
    },
  ],
};

const UserProvider = ({
  children,
  user = testUser,
  setToken = () => console.log("set token"),
}) => <UserContext.Provider value={{ user }}>{children}</UserContext.Provider>;

export { UserProvider };
