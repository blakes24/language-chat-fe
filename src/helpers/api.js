import axios from "axios";
import { BASE_URL } from "../config";
import { getFromLocalStorage } from "./localStorage";

/** API Class: methods used to connect to the API. */

let token = getFromLocalStorage("token", "");

class ChatApi {
  static token = token;

  static async request(endpoint, data = {}, method = "get") {
    console.debug("API Call:", endpoint, data, method);
    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${ChatApi.token}` };
    const params = method === "get" ? data : {};
    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      let message = err.response
        ? err.response.data.error.message
        : "Server error: try again later";
      throw Array.isArray(message) ? message : [message];
    }
  }

  /** Create a new user and get auth token. */

  static async register(userData) {
    let res = await this.request(`auth/register`, userData, "post");
    this.token = res.token;
    return res.token;
  }

  /** Get user details. */

  static async getUser(userId) {
    let res = await this.request(`users/${userId}`);
    return res.user;
  }

  /** Verifies that the social token provided is valid */

  static async verify(token) {
    let res = await this.request(`auth/validate`, token, "post");
    return res;
  }

  /** Get gets token to authenticate future requests to api */

  static async getToken(data) {
    let res = data.email
      ? await this.request(`auth/token`, data, "post")
      : await this.request(`auth/social-token`, data, "post");
    this.token = res.token;
    return res.token;
  }

  /** Get all users, filter by speaks or learning language */

  static async getAllUsers(filters) {
    let res = await this.request(`users`, filters);
    return res.users;
  }

  /** Get a users chat rooms, filter by chat partner */

  static async getRooms({ userId, partnerId }) {
    let queryParams = { partner: partnerId };
    let res = await this.request(`users/${userId}/rooms`, queryParams);
    return res;
  }

  /** Creates a chat room for two users data = {user1, user2} */

  static async createRoom(data) {
    let res = await this.request(`rooms`, data, "post");
    return res;
  }

  /** Gets messages, filter by room */

  static async getMessages(roomId) {
    let queryParams = { room: roomId };
    let res = await this.request(`messages`, queryParams);
    return res.messages;
  }

  /** Update user info. */

  static async updateUser(id, data) {
    let res = await this.request(`users/${id}`, data, "patch");
    return res.user;
  }

  /** Update language(speaks). */

  static async updateSpeaks(id, data) {
    let res = await this.request(`languages/speaks/${id}`, data, "patch");
    return res.language;
  }

  /** Update language(learning). */

  static async updateLearning(id, data) {
    let res = await this.request(`languages/learning/${id}`, data, "patch");
    return res.language;
  }

  /** Get a users partners */

  static async getPartners(userId) {
    let res = await this.request(`users/${userId}/partners`);
    return res.partners;
  }

  /** Adds a partner */

  static async addPartner(userId, data) {
    let res = await this.request(`users/${userId}/partners`, data, "post");
    return res;
  }

  /** Deletes a partner */

  static async deletePartner({ userId, partnerId }, data) {
    let res = await this.request(
      `users/${userId}/partners/${partnerId}`,
      data,
      "delete"
    );
    return res;
  }

  /** Deletes a user */

  static async deleteUser(userId, data) {
    let res = await this.request(`users/${userId}`, data, "delete");
    return res;
  }
}

export default ChatApi;
