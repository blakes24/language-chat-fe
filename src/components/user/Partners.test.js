import { render } from "@testing-library/react";
import Partners from "./Partners";
import { MemoryRouter } from "react-router-dom";
import { store } from "../../helpers/mockStore";
import { Provider } from "react-redux";

test("renders without crashing", () => {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <Partners />
      </MemoryRouter>
    </Provider>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <Provider store={store}>
      <MemoryRouter>
        <Partners />
      </MemoryRouter>
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});
