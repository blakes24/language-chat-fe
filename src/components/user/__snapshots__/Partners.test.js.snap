// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <h1
      class="MuiTypography-root makeStyles-root-5 MuiTypography-h4 MuiTypography-alignCenter"
    >
      Partners
    </h1>
    <p
      class="MuiTypography-root makeStyles-root-5 MuiTypography-body1 MuiTypography-alignCenter"
    >
      No partners yet. Find users to chat with 
      <a
        href="/"
      >
        here
      </a>
      .
    </p>
  </div>
</DocumentFragment>
`;
